chain/chain/src/validate.rs

```diff
@@ -189,30 +249,63 @@ fn validate_chunk_authorship(
 }
 
 fn validate_chunk_proofs_challenge(
+    chain_store: &mut ChainStore,
     runtime_adapter: &dyn RuntimeAdapter,
     chunk_proofs: &ChunkProofs,
+    transaction_validity_period: BlockIndex,
 ) -> Result<(CryptoHash, Vec<AccountId>), Error> {
     let block_header = BlockHeader::try_from_slice(&chunk_proofs.block_header)?;
     validate_header_authorship(runtime_adapter, &block_header)?;
-    let chunk_producer = validate_chunk_authorship(runtime_adapter, &chunk_proofs.chunk.header)?;
+    let chunk_header = match &chunk_proofs.chunk {
+        MaybeEncodedShardChunk::Encoded(encoded_chunk) => &encoded_chunk.header,
+        MaybeEncodedShardChunk::Decoded(chunk) => &chunk.header,
+    };
+    let chunk_producer = validate_chunk_authorship(runtime_adapter, &chunk_header)?;
     if !Block::validate_chunk_header_proof(
-        &chunk_proofs.chunk.header,
+        &chunk_header,
         &block_header.inner_rest.chunk_headers_root,
         &chunk_proofs.merkle_proof,
     ) {
+        // Merkle proof is invalid. It's a malicious challenge.
         return Err(ErrorKind::MaliciousChallenge.into());
     }
-    match chunk_proofs
-        .chunk
-        .decode_chunk(
-            runtime_adapter.num_data_parts(&chunk_proofs.chunk.header.inner.prev_block_hash),
-        )
-        .map_err(|err| err.into())
-        .and_then(|chunk| validate_chunk_proofs(&chunk, &*runtime_adapter))
-    {
-        Ok(true) => Err(ErrorKind::MaliciousChallenge.into()),
-        Ok(false) | Err(_) => Ok((block_header.hash(), vec![chunk_producer])),
+    // Temporary holds the decoded chunk, since we use a reference below to avoid cloning it.
+    let tmp_chunk;
+    let chunk_ref = match &chunk_proofs.chunk {
+        MaybeEncodedShardChunk::Encoded(encoded_chunk) => {
+            match encoded_chunk
+                .decode_chunk(runtime_adapter.num_data_parts(&chunk_header.inner.prev_block_hash))
+            {
+                Ok(chunk) => {
+                    tmp_chunk = Some(chunk);
+                    tmp_chunk.as_ref().unwrap()
+                }
+                Err(_) => {
+                    // Chunk can't be decoded. Good challenge.
+                    return Ok((block_header.hash(), vec![chunk_producer]));
```

Comment:
If the return type was a structure with a field `accounts_to_slash` or somesuch then it would be slightly easier to understand.

Reply:
Added a named variable for return type to avoid copying the return code multiple times.